File: 0415. Add Strings.java
// Easy
// Simulation?
// Time:O(n), Space:O(n)
// https://leetcode.cn/problems/add-strings/

class Solution {
  // Similar to Leetcode 2?
  public String addStrings(String num1, String num2) {
      StringBuilder sb = new StringBuilder();
      int carry = 0, i = num1.length() - 1, j = num2.length() - 1;
      while (i >= 0 || j >= 0 || carry != 0) {
          if (i >= 0) carry += num1.charAt(i--) - '0';
          if (j >= 0) carry += num2.charAt(j--) - '0';
          sb.append(carry % 10);
          carry /= 10;
      }
      return sb.reverse().toString();
  }
}

File: 0001-0500/0415. Add Strings.py
# Time:O(n), Space:O(n)

class Solution:
    def addStrings(self, num1: str, num2: str) -> str:
        res = []
        carry = 0
        i, j = len(num1) - 1, len(num2) - 1

        while i >= 0 or j >= 0 or carry:
            d1 = int(num1[i]) if i >= 0 else 0
            d2 = int(num2[j]) if j >= 0 else 0

            total = d1 + d2 + carry
            res.append(str(total % 10))
            carry = total // 10

            i -= 1
            j -= 1

        return "".join(reversed(res))
